File: app/modules/orbit/conversation.py
"""
Conversation Memory Manager for Orbit

Manages conversation history for contextual AI responses.
Uses Redis for fast, temporary storage with session-based isolation.
"""

import json
import logging
from typing import List, Dict, Optional
from datetime import datetime
from app.core.redis import redis_manager

logger = logging.getLogger(__name__)


class ConversationManager:
    """
    Manages conversation history in Redis
    
    Storage format:
    - Key: orbit:conversation:{user_id}:{session_id}
    - Value: JSON list of messages
    - TTL: 24 hours
    """
    
    def __init__(self, ttl_seconds: int = 86400):  # 24 hours default
        """Initialize conversation manager"""
        self.ttl_seconds = ttl_seconds
        self.key_prefix = "orbit:conversation"
    
    def _get_key(self, user_id: str, session_id: str) -> str:
        """Generate Redis key for conversation"""
        return f"{self.key_prefix}:{user_id}:{session_id}"
    
    def add_message(
        self,
        user_id: str,
        session_id: str,
        role: str,
        content: str
    ) -> bool:
        """
        Add message to conversation history
        
        Args:
            user_id: User UUID
            session_id: Session UUID
            role: 'user' or 'assistant'
            content: Message content
            
        Returns:
            True if successful
        """
        try:
            key = self._get_key(user_id, session_id)
            
            # Create message object
            message = {
                "role": role,
                "content": content,
                "timestamp": datetime.utcnow().isoformat()
            }
            
            # Get existing history
            history_json = redis_manager.get(key)
            
            if history_json:
                history = json.loads(history_json)
            else:
                history = []
            
            # Append new message
            history.append(message)
            
            # Keep only last 20 messages (10 turns) to save memory
            if len(history) > 20:
                history = history[-20:]
            
            # Save back to Redis with TTL
            redis_manager.setex(
                key,
                self.ttl_seconds,
                json.dumps(history)
            )
            
            logger.debug(f"Added {role} message to conversation {session_id}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to add message to conversation: {e}")
            return False
    
    def get_history(
        self,
        user_id: str,
        session_id: str,
        limit: Optional[int] = None
    ) -> List[Dict]:
        """
        Get conversation history
        
        Args:
            user_id: User UUID
            session_id: Session UUID
            limit: Maximum number of messages to return (most recent)
            
        Returns:
            List of messages in chronological order
        """
        try:
            key = self._get_key(user_id, session_id)
            history_json = redis_manager.get(key)
            
            if not history_json:
                return []
            
            history = json.loads(history_json)
            
            # Apply limit if specified
            if limit and len(history) > limit:
                history = history[-limit:]
            
            return history
            
        except Exception as e:
            logger.error(f"Failed to get conversation history: {e}")
            return []
    
    def clear_session(self, user_id: str, session_id: str) -> bool:
        """
        Clear conversation history for a session
        
        Args:
            user_id: User UUID
            session_id: Session UUID
            
        Returns:
            True if successful
        """
        try:
            key = self._get_key(user_id, session_id)
            redis_manager.delete(key)
            logger.info(f"Cleared conversation session {session_id}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to clear conversation: {e}")
            return False
    
    def get_message_count(self, user_id: str, session_id: str) -> int:
        """Get number of messages in conversation"""
        history = self.get_history(user_id, session_id)
        return len(history)
    
    def format_history_for_llm(
        self,
        user_id: str,
        session_id: str,
        limit: int = 10
    ) -> List[Dict[str, str]]:
        """
        Format conversation history for LLM consumption (OpenAI format)
        
        Returns:
            List of {"role": "user"/"assistant", "content": "..."}
        """
        history = self.get_history(user_id, session_id, limit)
        
        # Convert to OpenAI message format (remove timestamps)
        return [
            {"role": msg["role"], "content": msg["content"]}
            for msg in history
        ]


# Singleton instance
conversation_manager = ConversationManager()

File: app/modules/orbit-js/src/core/conversation.ts
/**
 * Conversation Memory Manager for Orbit
 * 
 * Manages conversation history for contextual AI responses.
 * Uses Redis for fast, temporary storage with session-based isolation.
 * Ported from Python conversation.py
 */

import Redis from 'ioredis';
import { config } from '../config';
import { ConversationMessage } from '../types';

/**
 * Conversation manager using Redis
 * 
 * Storage format:
 * - Key: orbit:conversation:{user_id}:{session_id}
 * - Value: JSON list of messages
 * - TTL: 24 hours (configurable)
 */
export class ConversationManager {
  private redis: Redis;
  private ttlSeconds: number;
  private maxMessages: number;
  private keyPrefix: string = 'orbit:conversation';

  constructor(redis?: Redis) {
    this.redis = redis || new Redis(config.redis.url, {
      retryStrategy: (times) => {
        const delay = Math.min(times * 50, 2000);
        return delay;
      },
      maxRetriesPerRequest: 3,
      lazyConnect: true,
    });
    
    this.ttlSeconds = config.conversation.ttlSeconds;
    this.maxMessages = config.conversation.maxMessages;
  }

  /**
   * Generate Redis key for conversation
   */
  private getKey(userId: string, sessionId: string): string {
    return `${this.keyPrefix}:${userId}:${sessionId}`;
  }

  /**
   * Add message to conversation history
   * 
   * @param userId - User UUID
   * @param sessionId - Session UUID
   * @param role - 'user' or 'assistant'
   * @param content - Message content
   * @returns True if successful
   */
  async addMessage(
    userId: string,
    sessionId: string,
    role: 'user' | 'assistant',
    content: string
  ): Promise<boolean> {
    try {
      const key = this.getKey(userId, sessionId);
      
      // Create message object
      const message: ConversationMessage = {
        role,
        content,
        timestamp: new Date().toISOString(),
      };

      // Get existing history
      const historyJson = await this.redis.get(key);
      let history: ConversationMessage[] = [];
      
      if (historyJson) {
        history = JSON.parse(historyJson);
      }

      // Append new message
      history.push(message);

      // Keep only last N messages to save memory
      if (history.length > this.maxMessages) {
        history = history.slice(-this.maxMessages);
      }

      // Save back to Redis with TTL
      await this.redis.setex(key, this.ttlSeconds, JSON.stringify(history));

      return true;
    } catch (error) {
      console.error('Failed to add message to conversation:', error);
      return false;
    }
  }

  /**
   * Get conversation history
   * 
   * @param userId - User UUID
   * @param sessionId - Session UUID
   * @param limit - Maximum number of messages to return (most recent)
   * @returns List of messages in chronological order
   */
  async getHistory(
    userId: string,
    sessionId: string,
    limit?: number
  ): Promise<ConversationMessage[]> {
    try {
      const key = this.getKey(userId, sessionId);
      const historyJson = await this.redis.get(key);

      if (!historyJson) {
        return [];
      }

      let history: ConversationMessage[] = JSON.parse(historyJson);

      // Apply limit if specified
      if (limit && history.length > limit) {
        history = history.slice(-limit);
      }

      return history;
    } catch (error) {
      console.error('Failed to get conversation history:', error);
      return [];
    }
  }

  /**
   * Clear conversation history for a session
   * 
   * @param userId - User UUID
   * @param sessionId - Session UUID
   * @returns True if successful
   */
  async clearSession(userId: string, sessionId: string): Promise<boolean> {
    try {
      const key = this.getKey(userId, sessionId);
      await this.redis.del(key);
      return true;
    } catch (error) {
      console.error('Failed to clear conversation:', error);
      return false;
    }
  }

  /**
   * Get number of messages in conversation
   */
  async getMessageCount(userId: string, sessionId: string): Promise<number> {
    const history = await this.getHistory(userId, sessionId);
    return history.length;
  }

  /**
   * Format conversation history for LLM consumption (OpenAI format)
   * 
   * Returns list of {role: "user"/"assistant", content: "..."}
   */
  async formatHistoryForLLM(
    userId: string,
    sessionId: string,
    limit: number = 10
  ): Promise<Array<{ role: 'user' | 'assistant'; content: string }>> {
    const history = await this.getHistory(userId, sessionId, limit);
    
    // Convert to OpenAI message format (remove timestamps)
    return history.map((msg) => ({
      role: msg.role,
      content: msg.content,
    }));
  }

  /**
   * Check if a session exists
   */
  async sessionExists(userId: string, sessionId: string): Promise<boolean> {
    const key = this.getKey(userId, sessionId);
    const exists = await this.redis.exists(key);
    return exists === 1;
  }

  /**
   * Get all active sessions for a user
   */
  async getUserSessions(userId: string): Promise<string[]> {
    try {
      const pattern = `${this.keyPrefix}:${userId}:*`;
      const keys = await this.redis.keys(pattern);
      
      // Extract session IDs from keys
      return keys.map((key) => {
        const parts = key.split(':');
        return parts[parts.length - 1];
      });
    } catch (error) {
      console.error('Failed to get user sessions:', error);
      return [];
    }
  }

  /**
   * Close Redis connection
   */
  async close(): Promise<void> {
    await this.redis.quit();
  }
}

// Singleton instance
let conversationManagerInstance: ConversationManager | null = null;

export function getConversationManager(): ConversationManager {
  if (!conversationManagerInstance) {
    conversationManagerInstance = new ConversationManager();
  }
  return conversationManagerInstance;
}
